<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <!-- Meta tags for social media banners, these should be filled in appropriatly as they are your "business card" -->
  <!-- Replace the content tag with appropriate information -->
  <meta name="description" content="DESCRIPTION META TAG">
  <meta property="og:title" content="SOCIAL MEDIA TITLE TAG"/>
  <meta property="og:description" content="SOCIAL MEDIA DESCRIPTION TAG TAG"/>
  <meta property="og:url" content="URL OF THE WEBSITE"/>
  <!-- Path to banner image, should be in the path listed below. Optimal dimenssions are 1200X630-->
  <meta property="og:image" content="static/image/your_banner_image.png" />
  <meta property="og:image:width" content="1200"/>
  <meta property="og:image:height" content="630"/>


  <meta name="twitter:title" content="TWITTER BANNER TITLE META TAG">
  <meta name="twitter:description" content="TWITTER BANNER DESCRIPTION META TAG">
  <!-- Path to banner image, should be in the path listed below. Optimal dimenssions are 1200X600-->
  <meta name="twitter:image" content="static/images/your_twitter_banner_image.png">
  <meta name="twitter:card" content="summary_large_image">
  <!-- Keywords for your paper to be indexed by-->
  <meta name="keywords" content="KEYWORDS SHOULD BE PLACED HERE">
  <meta name="viewport" content="width=device-width, initial-scale=1">


  <title>CodonTransformer</title>
  <link rel="icon" type="image/x-icon" href="static/images/favicon.ico">
  <link href="https://fonts.googleapis.com/css?family=Google+Sans|Noto+Sans|Castoro"
  rel="stylesheet">

  <link rel="stylesheet" href="static/css/bulma.min.css">
  <link rel="stylesheet" href="static/css/bulma-carousel.min.css">
  <link rel="stylesheet" href="static/css/bulma-slider.min.css">
  <link rel="stylesheet" href="static/css/fontawesome.all.min.css">
  <link rel="stylesheet"
  href="https://cdn.jsdelivr.net/gh/jpswalsh/academicons@1/css/academicons.min.css">
  <link rel="stylesheet" href="static/css/index.css">

  <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.5.1/jquery.min.js"></script>
  <script src="https://documentcloud.adobe.com/view-sdk/main.js"></script>
  <script defer src="static/js/fontawesome.all.min.js"></script>
  <script src="static/js/bulma-carousel.min.js"></script>
  <script src="static/js/bulma-slider.min.js"></script>
  <script src="static/js/index.js"></script>
</head>
<body>


  <section class="hero">
    <div class="hero-body">
      <div class="container is-max-desktop">
        <div class="columns is-centered">
          <div class="column has-text-centered">
            <div style="text-align:center; margin-bottom:20px;">
              <img src="static/images/logo.png" alt="CodonTransformer" style="vertical-align: middle; height: 100px; width: 100px; margin-right: 10px;">
              <h1 class="title is-1 publication-title" style="display: inline; vertical-align: middle;">CodonTransformer</h1>
            </div>          
              <h1 class="title is-2 publication-title">A Multispecies Codon Optimizer<br>Using Context-aware Neural Networks</h1>
                  <div class="is-size-5 publication-authors">
                    <!-- Paper authors -->
                    <span class="author-block">
                      <a href="https://adibvafa.github.io/Portfolio/" target="_blank">Adibvafa Fallahpour<sup>1, 2, 3*</sup>,&nbsp;</a>
                    </span>
                    <span class="author-block">
                      <a href="mailto:vincent.gureghian@free.fr" target="_blank">Vincent Gureghian<sup>1, 2*</sup>,&nbsp;</a>
                    </span>
                    <span class="author-block">
                      <a href="mailto:guillaume.filion@utoronto.ca" target="_blank">Guillaume J. Filion<sup>1†</sup>,&nbsp;</a>
                    </span>
                    <span class="author-block">
                      <a href="mailto:ariel.lindner@inserm.fr" target="_blank">Ariel B. Lindner<sup>2†</sup>,&nbsp;</a>
                    </span>
                    <span class="author-block">
                      <a href="mailto:amir.pandi@cri-paris.org" target="_blank">Amir Pandi<sup>2†</sup></a>
                    </span>
                  </div>
                  
                
                  <div class="is-size-5 publication-authors" style="margin-top:15px;">
                    <span class="author-block">
                      &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<sup>1</sup>
                      University of Toronto
                      <img src="static/images/uoft2.png" alt="University of Toronto" style="float:right;width:40px;height:40px;margin-top:-5px;">
                      </span>
                    <span class="author-block">
                      &nbsp;&nbsp;&nbsp;<sup>2</sup>
                      INSERM&nbsp;&nbsp;
                      <img src="static/images/inserm.png" alt="INSERM" style="float:right;width:35px;height:35px;margin-top:-1px;">
                      </span>
                    <span class="author-block">
                      &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<sup>3</sup>
                      Vector Institute&nbsp;
                      <img src="static/images/vector.png" alt="Vector Institute" style="float:right;width:30px;height:30px;">
                    </span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
                  </div>

                  <div class="author-notes", style="margin-top:15px;">
                    <p><sup>*</sup> Equal Contribution.&nbsp;&nbsp;&nbsp;<sup>†</sup> Equal Advising.</p>
                  </div>

                  <div class="column has-text-centered" style="margin-top: 5px;">
                    <div class="publication-links">
                         <!-- Arxiv PDF link -->
                      <span class="link-block">
                        <a href="https://arxiv.org/pdf/<ARXIV PAPER ID>.pdf" target="_blank"
                        class="external-link button is-normal is-rounded is-dark">
                        <span class="icon">
                          <i class="ai ai-arxiv"></i>
                        </span>
                        <span>Paper</span>
                      </a>
                    </span>

                    <!-- Supplementary PDF link -->
                    <!-- <span class="link-block">
                      <a href="static/pdfs/supplementary_material.pdf" target="_blank"
                      class="external-link button is-normal is-rounded is-dark">
                      <span class="icon">
                        <i class="fas fa-file-pdf"></i>
                      </span>
                      <span>Supplementary</span>
                    </a>
                  </span> -->

                  <!-- Github link -->
                  <span class="link-block">
                    <a href="https://github.com/Adibvafa/CodonTransformer" target="_blank"
                    class="external-link button is-normal is-rounded is-dark">
                    <span class="icon">
                      <i class="fab fa-github"></i>
                    </span>
                    <span>Code</span>
                  </a>
                </span>

                <!-- ArXiv abstract Link -->
                <!-- <span class="link-block">
                  <a href="https://arxiv.org/abs/<ARXIV PAPER ID>" target="_blank"
                  class="external-link button is-normal is-rounded is-dark">
                  <span class="icon">
                    <i class="ai ai-arxiv"></i>
                  </span>
                  <span>arXiv</span>
                </a>
              </span> -->

              <!-- Data link -->
              <span class="link-block">
                <a href="https://huggingface.co/datasets/adibvafa/CodonTransformer" target="_blank" class="external-link button is-normal is-rounded is-dark">
                  <span class="icon">
                    <i class="fas fa-database"></i>
                  </span>
                  <span>Data</span>
                </a>
              </span>

              <!-- HuggingFace link -->
              <span class="link-block">
                <a href="https://huggingface.co/adibvafa/CodonTransformer" target="_blank" class="external-link button is-normal is-rounded is-dark">
                  <span class="icon">
                    <p>&#129303;</p>
                  </span>
                  <span>Model</span>
                </a>
              </span>

              <!-- Demo link -->
              <span class="link-block">
                <a href="https://adibvafa.github.io/CodonTransformer/GoogleColab" target="_blank" class="external-link button is-normal is-rounded is-dark">
                  <span class="icon">
                    <img src="static/images/colab.png" alt="Colab Logo" style="width: 20px; height: 20px;"> <!-- Adjust the path as needed -->
                  </span>
                  <span>Colab Notebook</span>
                </a>
              </span>

            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</section>

<!-- 
<div style="text-align:center; margin-bottom:50px;">
  <img src="static/images/codontransformer.png" alt="CodonTransformer" style="vertical-align: middle; width:20%; margin-right: 10px;">
</div> -->

<!-- Paper abstract -->
<!-- <section class="section hero light">
  <div class="container is-max-widescreen">
    <div class="columns is-centered has-text-centered">
      <div class="column is-nine-tenth">
        <h2 class="title is-3">Abstract</h2>
        <div class="content has-text-justified is-size-abstract">
          <p>
            The genetic code is degenerate allowing a multitude of possible DNA sequences to encode the same protein. This degeneracy impacts the efficiency of heterologous protein production due to the codon usage preferences of each organism. The process of tailoring organism-specific synonymous codons, known as codon optimization, must respect local sequence patterns that go beyond global codon preferences. As a result, the search space faces a combinatorial explosion that makes exhaustive exploration impossible. Nevertheless, throughout the diverse life on Earth, natural selection has already optimized the sequences, thereby providing a rich source of data allowing machine learning algorithms to explore the underlying rules. Here, we introduce CodonTransformer, a multispecies deep learning model trained on over 1 million DNA-protein pairs from 164 organisms spanning all kingdoms of life. The model demonstrates context-awareness thanks to the attention mechanism and bidirectionality of the Transformers we used, and to a novel sequence representation that combines organism, amino acid, and codon encodings. CodonTransformer generates host-specific DNA sequences with natural-like codon distribution profiles and with negative cis-regulatory elements. This work introduces a novel strategy of Shared Token Representation and Encoding with Aligned Multi-masking (STREAM) and provides a state-of-the-art codon optimization framework with a customizable open-access model and a user-friendly interface.
          </p>
        </div>
      </div>
    </div>
  </div>
</section> -->
<!-- End paper abstract -->

<section class="section" id="Home">
  <div class="container is-max-desktop">
    <h2 class="title is-3">Home</h2>
    <div class="content is-size-5 has-text-justified">
      <p> Welcome to the CodonTransformer project page!       </p>
      <p>
        <strong>CodonTransformer</strong> is a cutting-edge, multispecies deep learning model designed for state-of-the-art codon optimization. Trained on over 1 million gene-protein pairs from 164 organisms spanning all kingdoms of life, CodonTransformer leverages advanced neural network architectures to generate host-specific DNA sequences with natural-like codon usage patterns and minimized negative cis-regulatory elements for any protein sequence.      </p>
    </div>
  </div>

  <div style="display: flex; justify-content: center; align-items: center; margin-top: 50px;">
    <a href="https://adibvafa.github.io/CodonTransformer/GoogleColab" target="_blank" class="button is-rounded is-dark" style="display: flex; align-items: center; font-size: 1.25rem; padding: 1.6rem 2rem;">
      <img src="static/images/colab.png" alt="Colab Logo" style="width: 40px; height: 40px; margin-right: 20px;">
      <span>Try CodonTransformer Now!</span>
    </a>
  </div>

</section>

<section class="section" id="Overview">
  <div class="container is-max-desktop">
    <h2 class="title is-3">Overview</h2>
    <div class="content is-size-5 has-text-justified">
      <p> The genetic code's degeneracy allows for multiple DNA sequences to encode the same protein, but not all codons are equal in the eyes of the host organism. Codon usage bias can significantly impact the efficiency of heterologous protein production due to differences in tRNA abundance, protein folding regulations, and evolutionary constraints.</p>
      <p> CodonTransformer addresses this challenge by using the Transformer architecture and a novel training strategy named STREAM to learn and replicate the intricate codon usage patterns across a diverse array of organisms. By doing so, it provides a powerful tool for tailoring DNA sequences to optimize protein expression in various host species. </p>  
    </div>
    <div class="columns is-centered" style="margin-top:15px;text-align:center;">
      <div class="column is-nine-tenth">
        <img src="static/images/model.png" alt="Data" style="width:65%;">
      </div>
    </div>
  </div>
</section>

<section class="section" id="Methods">
  <div class="container is-max-desktop">
    <h2 class="title is-3">Methods</h2>
    <div class="content is-size-5 has-text-justified">
      <p>
        CodonTransformer utilizes the BigBird Transformer architecture, an advanced neural network model capable of efficiently processing long sequences. A novel sequence representation strategy, Shared Token Representation and Encoding with Aligned Multi-masking (STREAM), combines organism encoding with tokenized amino acid-codon pairs to achieve context-awareness.

        This innovative approach allows the model to capture both global and local codon usage patterns, enabling it to generate DNA sequences that are finely tuned to the host organism's preferences.
      </p>
      <p><strong>Key Features</strong></p>
      <ul>
        <li><strong>Encoder-Only:</strong> Bidirectional design allows uniform optimization across the entire sequence.</li>
        <li><strong>STREAM Encoding:</strong> Integrates organism, amino acid, and codon information into shared tokens.</li>
        <li><strong>Masked Language Modeling:</strong> Learning complex patterns by predicting masked codons.</li>
      </ul>
    </div>
    <!-- <div class="columns is-centered" style="margin-top:15px;text-align:center;">
      <div class="column is-nine-tenth">
        <img src="static/images/model.png" alt="Data" style="width:65%;">
      </div>
    </div>
  </div> -->
</section>

<section class="section" id="Results">
  <div class="container is-max-desktop">
    <h2 class="title is-3">CodonTransformer Outperforms Existing Tools</h2>
    <div class="content is-size-5 has-text-justified">
      <p>
        Through extensive testing and benchmarking, CodonTransformer demonstrates superior performance in generating natural-like codon distributions and minimizing negative cis-regulatory elements compared to existing codon optimization tools.
      </p>
    </div>
  </div>
</section>

<section class="section" id="Results2">
  <div class="container is-max-desktop">
    <h2 class="title is-3">Learning Codon Patterns Across Organisms</h2>
    <div class="content is-size-5 has-text-justified">
      <p>
        CodonTransformer effectively learned codon usage patterns across multiple species, as shown by high codon similarity indices (CSI) when generating DNA sequences for various organisms. The model adapts to the specific codon preferences of each host, ensuring optimal expression.
      </p>
    </div>
    <div class="columns is-centered" style="margin-top:15px;text-align:center;">
      <div class="column is-nine-tenth">
        <img src="static/images/Fig. 2.jpg" alt="Data" style="width:95%;">
      </div>
    </div>
  </div>
</section>

<section class="section" id="Results3">
  <div class="container is-max-desktop">
    <h2 class="title is-3">Generating Natural-Like Codon Distributions</h2>
    <div class="content is-size-5 has-text-justified">
      <p>
        The model produces DNA sequences with codon usage patterns closely resembling those found in nature, avoiding clusters of rare or highly frequent codons that can negatively affect protein folding and expression. This is visualized using %MinMax profiles and Dynamic Time Warping (DTW) distance metrics.
        </p>
    </div>
    <div class="columns is-centered" style="margin-top:15px;text-align:center;">
      <div class="column is-nine-tenth">
        <img src="static/images/Fig. 3.jpg" alt="Data" style="width:95%;">
      </div>
    </div>
  </div>
</section>

<section class="section" id="Results4">
  <div class="container is-max-desktop">
    <h2 class="title is-3">Benchmarking with Real World Proteins</h2>
    <div class="content is-size-5 has-text-justified">
      <p>
        When benchmarked against proteins of biotechnological interest, CodonTransformer consistently generates sequences with minimized negative cis-regulatory elements, outperforming other tools. This enhances the potential for successful heterologous expression.
      </p>
    </div>
    <div class="columns is-centered" style="margin-top:15px;text-align:center;">
      <div class="column is-nine-tenth">
        <img src="static/images/Fig. 4.jpg" alt="Data" style="width:95%;">
      </div>
    </div>
  </div>
</section>

<section class="section" id="Features">
  <div class="container is-max-desktop">
    <h2 class="title is-3">Why Choose CodonTransformer?</h2>
    <div class="content is-size-5 has-text-justified">
      <ul>
        <li><strong>Multispecies Support:</strong> Trained on 164 organisms, CodonTransformer can optimize codon usage for a wide range of host species, including prokaryotes and eukaryotes.</li>
        <li><strong>Context-Aware Optimization:</strong> The model considers both global codon usage biases and local sequence patterns, ensuring optimal DNA sequence design.</li>
        <li><strong>Natural-Like Codon Distribution:</strong> Generates sequences with codon distributions similar to natural genes, aiding in proper protein folding and function.</li>
        <li><strong>Minimizing Negative Cis-Regulatory Elements:</strong> Avoids sequences that could interfere with gene expression, enhancing the success rate of protein production.</li>
        <li><strong>Open-Access and Customizable:</strong> The base and fine-tuned models are openly available, along with a comprehensive Python package and a user-friendly Google Colab notebook.</li>
        <li><strong>Custom Fine-Tuning:</strong> Users can fine-tune the model on custom datasets to meet specific requirements or optimize for unique organisms.</li>
      </ul>
    </div>
  </div>
</section>


<!--BibTex citation -->
  <section class="section" id="BibTeX">
    <div class="container is-max-desktop content">
      <h2 class="title">BibTeX</h2>
      <pre><code>BibTex Code Here</code></pre>
    </div>
</section>
<!--End BibTex citation -->


  <footer class="footer">
  <div class="container">
    <div class="columns is-centered">
      <div class="column is-8">
        <div class="content">

          <p>
            This page was built using the <a href="https://github.com/eliahuhorwitz/Academic-project-page-template" target="_blank">Academic Project Page Template</a> which was adopted from the <a href="https://nerfies.github.io" target="_blank">Nerfies</a> project page.
            You are free to borrow the of this website, we just ask that you link back to this page in the footer. <br> This website is licensed under a <a rel="license"  href="http://creativecommons.org/licenses/by-sa/4.0/" target="_blank">Creative
            Commons Attribution-ShareAlike 4.0 International License</a>.
          </p>

        </div>
      </div>
    </div>
  </div>
</footer>

<!-- Statcounter tracking code -->
  
<!-- You can add a tracker to track page visits by creating an account at statcounter.com -->

    <!-- End of Statcounter Code -->

  </body>
  </html>
